src/cdi_CPEloss/Tabular_CP_Eloss.cc

```diff
@@ -0,0 +1,270 @@
+      static_cast<int>(std::ceil(static_cast<double>(n_energy) / max_entries));
+  line_entries = read_line();
+  min_log_energy = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  energies.resize(n_energy);
+  for (uint32_t n = 0; n < n_energy; n++) {
+    energies[n] = exp(min_log_energy + n * d_log_energy);
+  }
+
+  // Get first density support point
+  nlines =
+      static_cast<int>(std::ceil(static_cast<double>(n_density) / max_entries));
+  line_entries = read_line();
+  min_log_density = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  densities.resize(n_density);
+  for (uint32_t n = 0; n < n_density; n++) {
+    densities[n] = exp(min_log_density + n * d_log_density);
+  }
+
+  // Get first temperature support point
+  nlines = static_cast<int>(
+      std::ceil(static_cast<double>(n_temperature) / max_entries));
+  line_entries = read_line();
+  min_log_temperature = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  temperatures.resize(n_temperature);
+  for (uint32_t n = 0; n < n_temperature; n++) {
+    temperatures[n] = exp(min_log_temperature + n * d_log_temperature);
+  }
+
+  std::vector<double> stopping_data_1d(n_energy * n_density * n_temperature);
+
+  bool target_found = false;
+  nlines = static_cast<int>(std::ceil(
+      ((double)n_energy * n_density * n_temperature) /
+      max_entries)); // The number of lines taken up by stopping power data for one target
+  if (target.get_zaid() == -1) {
+    // Target is free electrons
+    target_found = true;
+    uint32_t nentry = 0;
+    for (uint32_t n = 0; n < nlines; n++) {
+      line_entries = read_line();
+      for (std::string entry : line_entries) {
+        stopping_data_1d[nentry] = stod(entry);
+        nentry++;
+      }
+    }
+  } else {
+    // Skip electrons
+    for (uint32_t n = 0; n < nlines; n++) {
+      line_entries = read_line();
+    }
+
+    // Find ion target, if it exists
+    const uint32_t n_target_ions =
+        stoi(read_line()[0]); // Number of target ions in file
+    for (uint32_t n_target_ion = 0; n_target_ion < n_target_ions;
+         n_target_ion++) {
+      int zaid_target_ion = stoi(read_line()[0]); // ZAID
+      read_line();                                // Z, A, mass
+      if (zaid_target_ion == target.get_zaid()) {
+        // This is the requested target ion
+        target_found = true;
+        uint32_t nentry = 0;
+        for (uint32_t n = 0; n < nlines; n++) {
+          line_entries = read_line();
+          for (std::string entry : line_entries) {
+            stopping_data_1d[nentry] = stod(entry);
+            nentry++;
+          }
+        }
+        break;
+      } else {
+        // This is not the requested target ion
+        for (uint32_t n = 0; n < nlines; n++) {
+          read_line();
+        }
+      }
+    }
+  }
+  file.close();
+
+  std::ostringstream dedx_error;
```

Comment:
You might want to avoid creating this ostringstream object as I mentioned above.